### PR TITLE
feat: clickable links tracking

### LINKED ISSUE #256: Track which links are clicked in each page
Since we already have clickable links implemented (see #236), we should add functionality to track which links are clicked on each page. This will help us understand user engagement and which links are most valuable to users.

**Context:**
Following up on #236 which added support for clickable links in PDFs, we should now implement analytics/tracking to monitor:
- Which links are clicked
- How many times each link is clicked
- Which pages have the most link interactions

**Goals:**
- Implement link click tracking
- Record which links are clicked per page
- Enable analytics/insights on link usage patterns

**Acceptance Criteria:**
- Link clicks are tracked when users interact with clickable links
- Click data is stored and can be retrieved per page
- Analytics can show which links are most frequently used

### PR DESCRIPTION
**Pull request by xiez**

### CODE REVIEW

File: backend/sharelinks/serializers.py [hunk 1]
```diff
@@ -187,6 +197,40 @@ def create(self, validated_data):
         return super().create(validated_data)
 
 
+class LinkClickRecordSerializer(serializers.ModelSerializer):
+    """
+    Serializer for validating and creating LinkClick records from tracking data.
+    """
+    view_session = serializers.PrimaryKeyRelatedField(
+        queryset=ViewSession.objects.all()
```
Comment (xiez) [inline backend/sharelinks/serializers.py:206]: 🔴 **Must Fix — No ownership scoping on `view_session` (analytics poisoning)**

This endpoint is `AllowAny`, and `view_session` is looked up from the entire `ViewSession` table with no scoping. An attacker who observes any `view_session` ULID from a public share link can POST arbitrary `url` + `page_number` payloads attributed to that session, polluting the analytics dashboard of another organization.

This is more severe than the equivalent gap in `RecordPageView` because the stored `url` string is then rendered (even if safely) in another user's analytics view.

**Fix:** At minimum, tie the lookup to sessions belonging to share links that are currently accessible. Consider accepting the share link `slug` in the payload and scoping the session lookup to `ViewSession.objects.filter(share_link__slug=slug)`.

File: backend/sharelinks/serializers.py [hunk 2]
```diff
@@ -187,6 +197,40 @@ def create(self, validated_data):
...
+        model = LinkClick
+        fields = ['view_session', 'dataroom_visit', 'url', 'page_number']
+
+    def validate(self, data):
+        view_session = data.get('view_session')
+        dataroom_visit = data.get('dataroom_visit')
+        url = data.get('url')
+
+        if url:
+            if not (url.startswith('http://') or url.startswith('https://')):
```
Comment (xiez) [inline backend/sharelinks/serializers.py]: 🔴 **Must Fix — URL validation uses raw `startswith` with no stripping**

The check `url.startswith('http://')` runs against the raw, untrimmed input. A payload like `"  javascript:alert(1)"` (leading spaces) or `"\x00https://evil.com"` (leading null byte) passes this validation and gets persisted to the database.

The frontend `isSafeUrl` does strip control chars before rendering, so this isn’t directly exploitable in the current UI — but the raw dirty value is stored in the DB and could be dangerous if the display logic ever changes.

**Fix:** Mirror the `isSafeUrl` stripping logic on the backend before the check:
```python
import re
url = re.sub(r'[\x00-\x1F\x7F-\x9F]', '', url.strip())
if not (url.startswith('http://') or url.startswith('https://')):
    raise serializers.ValidationError(...)
```

File: backend/sharelinks/serializers.py [hunk 3]
```diff
@@ -42,17 +42,26 @@ def get_url(self, obj) -> str | None:
         return pages_map.get(obj.page_number)
 
 
+class LinkClickSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = LinkClick
+        fields = ['id', 'view_session', 'dataroom_visit', 'url', 'page_number', 'clicked_at']
```
Comment (xiez) [inline backend/sharelinks/serializers.py]: 🟡 **Should Fix — `LinkClickSerializer` exposes redundant FK IDs**

`view_session` and `dataroom_visit` are included in the serializer fields. When this serializer is nested inside `ViewSessionSerializer` and `DataroomVisitSerializer`, those FK IDs are completely redundant (the parent object already provides the context) and unnecessarily expose internal identifiers.

**Fix:** Trim to only what the UI actually consumes:
```python
fields = ['id', 'url', 'page_number', 'clicked_at']
```

File: frontend/src/components/documents/PdfJsViewer.jsx [hunk 4]
```diff
@@ -70,6 +70,19 @@ export const PdfJsViewer = forwardRef(({
...
+    if (!viewId) return;
+    const payload = {
+      view_session: viewId,
+      page_number: pageNumber,
+      url: url,
+    };
+    if (dataroomVisitId) {
+      payload.dataroom_visit = dataroomVisitId;
+    }
+    recordLinkClick(payload);
```
Comment (xiez) [inline frontend/src/components/documents/PdfJsViewer.jsx]: 🟡 **Should Fix — Fire-and-forget Promise with no `.catch()`**

`recordLinkClick(payload)` returns a Promise but it is neither awaited nor caught here. Since the `<a>` opens in a new tab (`target="_blank"`), navigation isn’t blocked — but a network failure will produce a silent unhandled rejection.

**Fix:** Suppress explicitly:
```js
recordLinkClick(payload).catch(() => {});
```

File: frontend/src/components/documents/PreviewViewer.jsx [hunk 5]
```diff
@@ -40,6 +40,19 @@ export const PreviewViewer = forwardRef(({ documentData, zoomLevel, onPageChange
...
+    if (!viewId) return;
+    const payload = {
+      view_session: viewId,
+      page_number: pageNumber,
+      url: url,
+    };
+    if (dataroomVisitId) {
+      payload.dataroom_visit = dataroomVisitId;
+    }
+    recordLinkClick(payload);
```
Comment (xiez) [inline frontend/src/components/documents/PreviewViewer.jsx]: 🟡 **Should Fix — Duplicated `handleLinkClick` implementation**

This function is **identical** to the one in `PdfJsViewer.jsx` (same payload shape, same `dataroomVisitId` branching, same `recordLinkClick` call). This copy-paste will silently drift if the telemetry payload ever changes.

**Fix:** Extract into a shared custom hook:
```js
// hooks/useLinkClickTracking.js
export function useLinkClickTracking(viewId, dataroomVisitId) {
  return useCallback((url, pageNumber) => {
    if (!viewId) return;
    const payload = { view_session: viewId, page_number: pageNumber, url };
    if (dataroomVisitId) payload.dataroom_visit = dataroomVisitId;
    recordLinkClick(payload).catch(() => {});
  }, [viewId, dataroomVisitId]);
}
```

File: backend/tests/sharelinks/test_views.py [hunk 6]
```diff
@@ -215,6 +215,65 @@ def test_record_page_view_with_mismatched_dataroom_visit_fails(self, public_clie
...
+
+        data = {
+            'view_session': view_session.id,
+            'url': 'javascript:alert(1)',
+            'page_number': 1
+        }
+        response = public_client.post('/api/v1/link-clicks/record/', data)
+        assert response.status_code == status.HTTP_400_BAD_REQUEST
+        assert 'url' in response.data
+        assert "Only outbound HTTP and HTTPS URLs are allowed" in response.data['url'][0]
```
Comment (xiez) [inline backend/tests/sharelinks/test_views.py:272]: 🟡 **Should Fix — Missing edge-case URL tests**

The test suite covers `javascript:alert(1)` being rejected and `https://google.com` accepted. Missing:

1. **Whitespace bypass** — `"  javascript:alert(1)"` (leading spaces): currently passes validation since `startswith` is not run against a stripped value.
2. **`data:` URI** — worth an explicit rejection test.
3. **`http://` success path** — only `https://` is tested as an accepted protocol.

```python
def test_record_link_click_whitespace_bypass(self, public_client, share_link):
    view_session = ViewSession.objects.create(share_link=share_link)
    data = {'view_session': view_session.id, 'url': '  javascript:alert(1)', 'page_number': 1}
    response = public_client.post('/api/v1/link-clicks/record/', data)
    assert response.status_code == status.HTTP_400_BAD_REQUEST
    assert LinkClick.objects.count() == 0
```

**Review by xiez — COMMENTED** [review]
## Code Review — `215ac1e` — feat: clickable links tracking

### ✅ Positives
- `isSafeUrl` correctly guards link overlays in both viewers *and* the analytics dashboard — the XSS threads flagged by the Gemini bot are already handled.
- `on_delete=SET_NULL` on `dataroom_visit` in `models.py` is correct and consistent with the migration.
- `keepalive: true` + `sendBeacon` fallback in `recordLinkClick` matches the project’s telemetry reliability policy.
- `prefetch_related` added in both `ShareLinkViewSet.view_sessions` and `ViewSessionViewSet.get_queryset` — N+1 is avoided.
- Backend URL validation present via `LinkClickRecordSerializer.validate()`.

### Issues
See inline comments. Two must-fix security/integrity issues and four should-fix quality items.
